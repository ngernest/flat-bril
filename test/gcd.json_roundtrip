{
  "functions": [
    {
      "args": [
        {
          "name": "op1",
          "type": "int"
        },
        {
          "name": "op2",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "vc0",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "op1"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "op2"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "cmp.val"
        },
        {
          "args": [
            "v0",
            "v1"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "v2"
          ],
          "funcs": [],
          "labels": [
            "if.1",
            "else.1"
          ],
          "op": "br"
        },
        {
          "label": "if.1"
        },
        {
          "args": [
            "v1",
            "v0"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "loop.bound"
          ],
          "op": "jmp"
        },
        {
          "label": "else.1"
        },
        {
          "args": [
            "v0",
            "v1"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "loop.bound"
          ],
          "op": "jmp"
        },
        {
          "label": "loop.bound"
        },
        {
          "args": [
            "v3",
            "vc0"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [
            "program.end",
            "update.val"
          ],
          "op": "br"
        },
        {
          "label": "update.val"
        },
        {
          "args": [
            "v2"
          ],
          "funcs": [],
          "labels": [
            "if.2",
            "else.2"
          ],
          "op": "br"
        },
        {
          "label": "if.2"
        },
        {
          "args": [
            "v3"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "cmp.val"
          ],
          "op": "jmp"
        },
        {
          "label": "else.2"
        },
        {
          "args": [
            "v3"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "cmp.val"
          ],
          "op": "jmp"
        },
        {
          "label": "program.end"
        },
        {
          "args": [
            "v1"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
