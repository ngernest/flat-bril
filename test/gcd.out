@main
	op: const		dest: vc0		type: int		value: 0	
	op: id   		dest: v0		type: int		args: ["op1"]	
	op: id   		dest: v1		type: int		args: ["op2"]	
.cmp.val
	op: lt   		dest: v2		type: bool		args: ["v0", "v1"]	
	op: br   		args: ["v2"]		labels: ["if.1", "else.1"]
.if.1
	op: sub  		dest: v3		type: int		args: ["v1", "v0"]	
	op: jmp  		labels: ["loop.bound"]
.else.1
	op: sub  		dest: v3		type: int		args: ["v0", "v1"]	
	op: jmp  		labels: ["loop.bound"]
.loop.bound
	op: eq   		dest: v4		type: bool		args: ["v3", "vc0"]	
	op: br   		args: ["v4"]		labels: ["program.end", "update.val"]
.update.val
	op: br   		args: ["v2"]		labels: ["if.2", "else.2"]
.if.2
	op: id   		dest: v1		type: int		args: ["v3"]	
	op: jmp  		labels: ["cmp.val"]
.else.2
	op: id   		dest: v0		type: int		args: ["v3"]	
	op: jmp  		labels: ["cmp.val"]
.program.end
	op: print		args: ["v1"]	
